homeassistant/components/powerwall/switch.py

```diff
@@ -0,0 +1,74 @@
+"""Support for Powerwall Switches (V2 API only)."""
+
+from typing import Any
+
+from tesla_powerwall import GridStatus, IslandMode, Powerwall
+
+from homeassistant.components.switch import SwitchDeviceClass, SwitchEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity import EntityCategory
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import DOMAIN, POWERWALL_API, POWERWALL_COORDINATOR
+from .models import PowerwallRuntimeData
+
+ICON = "mdi:home-battery-outline"
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up Powerwall switch platform from Powerwall resources."""
+    powerwall_data: PowerwallRuntimeData = hass.data[DOMAIN][config_entry.entry_id]
+    power_wall = powerwall_data[POWERWALL_API]
+    coordinator = powerwall_data[POWERWALL_COORDINATOR]
+    assert coordinator is not None
+
+    async_add_entities([PowerwallOffGridEnabledEntity(power_wall, coordinator)])
+
+
+class PowerwallOffGridEnabledEntity(SwitchEntity):
+    """Representation of a Switch entity for Powerwall Off-grid operation."""
+
+    _attr_name = "Take Powerwall Off-Grid"
+    _attr_unique_id = "take_powerwall_off_grid"
+    _attr_entity_category = EntityCategory.CONFIG
+    _attr_device_class = SwitchDeviceClass.SWITCH
+    _attr_icon = ICON
+
+    def __init__(
+        self,
+        power_wall: Powerwall,
+        coordinator: DataUpdateCoordinator,
+    ) -> None:
+        """Initialise the entity."""
+        self.power_wall = power_wall
+        self.coordinator = coordinator
+
+    @property
+    def is_on(self) -> bool:
+        """Return true if the powerwall is off-grid."""
+        return self.coordinator.data.grid_status in [
+            GridStatus.TRANSITION_TO_ISLAND,
+            GridStatus.ISLANDED,
+        ]
+
+    async def async_turn_on(self, **kwargs: Any) -> None:
+        """Turn off-grid mode on."""
+        self._attr_is_on = True
+        await self.hass.async_add_executor_job(
+            self.power_wall.set_island_mode, IslandMode.OFFGRID
+        )
```

Comment:
If the `set_island_mode` can fail it should raise `HomeAssistantError`

Operations like service calls and entity methods (e.g. Set HVAC Mode) have proper exception handling. Raise ValueError on invalid user input and raise HomeAssistantError for other failures such as a problem communicating with a device.

We shouldn't change the on state unless we know the call was successful so we set the attribute after the call has not failed and than write state

```suggestion
        await self.hass.async_add_executor_job(
            self.power_wall.set_island_mode, IslandMode.OFFGRID
        )
        self._attr_is_on = True
        self.async_write_ha_state()
```